Composer/packages/tools/language-servers/language-generation/src/LGServer.ts

```diff
@@ -0,0 +1,355 @@
+  protected workspaceRoot: URI | undefined;
+  protected readonly documents = new TextDocuments();
+  protected readonly pendingValidationRequests = new Map<string, number>();
+  protected LGDocuments: LGDocument[] = []; // LG Documents Store
+
+  constructor(protected readonly connection: IConnection) {
+    this.documents.listen(this.connection);
+    this.documents.onDidChangeContent(change => this.validate(change.document));
+    this.documents.onDidClose(event => {
+      this.cleanPendingValidation(event.document);
+      this.cleanDiagnostics(event.document);
+    });
+
+    this.connection.onInitialize(params => {
+      if (params.rootPath) {
+        this.workspaceRoot = URI.file(params.rootPath);
+      } else if (params.rootUri) {
+        this.workspaceRoot = URI.parse(params.rootUri);
+      }
+      this.connection.console.log('The server is initialized.');
+      return {
+        capabilities: {
+          textDocumentSync: this.documents.syncKind,
+          codeActionProvider: false,
+          completionProvider: {
+            resolveProvider: true,
+          },
+          hoverProvider: true,
+          foldingRangeProvider: false,
+        },
+      };
+    });
+    this.connection.onCompletion(params => this.completion(params));
+    this.connection.onHover(params => this.hover(params));
+
+    this.connection.onRequest((method, params) => {
+      if (InitializeDocumentsMethodName === method) {
+        const { uri, content, template } = params;
+        this.LGDocuments.push({ uri, content, template });
+        // run diagnostic
+        const textDocument = this.documents.get(uri);
+        if (textDocument) {
+          setTimeout(() => {
+            this.validate(textDocument);
+          }, 500);
+        }
+      }
+    });
+  }
+
+  start() {
+    this.connection.listen();
+  }
+
+  protected getLGDocument(document: TextDocument): LGDocument | undefined {
+    const LGDocument = this.LGDocuments.find(item => item.uri === document.uri);
+    return LGDocument;
+  }
+  protected getLGDocumentContent(document: TextDocument): string {
+    const LGDocument = this.LGDocuments.find(item => item.uri === document.uri);
+    let text = document.getText();
+    if (LGDocument) {
+      // concat new content
+      const { content, template } = LGDocument;
+      try {
+        text = updateTemplateInContent(content, template);
+        // eslint-disable-next-line no-empty
+      } catch (error) {}
+    }
+    return text;
+  }
+
+  protected hover(params: TextDocumentPositionParams): Thenable<Hover | null> {
+    const document = this.documents.get(params.textDocument.uri);
+    if (!document) {
+      return Promise.resolve(null);
+    }
+    const text = this.getLGDocumentContent(document);
+    let templates: LGTemplate[] = [];
+    try {
+      const lgResources = getLGResources(text);
+      templates = lgResources.Templates;
+      // eslint-disable-next-line no-empty
+    } catch (error) {}
```

Comment:
Try not to silence errors.

In this case, if LG document somehow corrupted or the parser has coding errors. `templates` will be empty. But in fact there should be content inside.

That means, this could accidentally destroy user data.

Reply:
refactor it, found trycatch is not necessary